nemo/collections/nlp/models/nlp_model.py

```diff
@@ -57,9 +58,17 @@ def __init__(self, cfg: DictConfig, trainer: Trainer = None):
         super().__init__(cfg, trainer)
         self.set_world_size(trainer)
 
-    def register_artifact(self, config_path: str, src: str, verify_src_exists: bool = False):
+    def register_artifact(
+        self,
+        config_path: str,
+        src: str,
+        verify_src_exists: bool = False,
+        save_restore_connector: SaveRestoreConnector = SaveRestoreConnector(),
```

Comment:
All function arguments should generally be immutable, I'd rather this save restore default argument be None, and in the implementation, None be replaced by an object of SaveRestoreConnecter. Please update this throughout all such calls. Such a check and replace can be done inside the update save restore connector method itself 

Reason: function arguments are parsed exactly once - at function definition time, not at function runtime. https://stackoverflow.com/questions/1132941/least-astonishment-and-the-mutable-default-argument

Reply:
+1 for Som's comment on immutable args.

But aside of it I need to go deeper into implementation of ModelPT, but for now not fully understanding the logic here: why do we have to add connector to `register_artifacts`? Do we assume that the connector for saving/restoring of a model can change for many times once a process runs, e.g. each time a different artifact is added?

Reply:
Re: the connector for register_artifact, it's only there because in NLPModel we don't call super.__init__ until after setting up tokenizers. If we can fix that, then we can remove connector from register_artifact.

Reply:
Defaulted all save_restore_connector args to None.